core/src/main/java/com/microsoft/applicationinsights/internal/util/PeriodicTaskManager.java

```diff
@@ -0,0 +1,260 @@
+ *
+ * <pre> {@code
+ * import com.microsoft.applicationinsights.internal.util.PeriodicTaskManager;
+ * public class BeeperControl {
+ *     public void beepForHour() {
+ *        final Runnable beeper = new Runnable() {
+ *            public void run() { System.out.println("beep"); }
+ *         };
+ *
+ *         // create the PeriodicTask from runnable
+ *         PeriodicTaskManager.PeriodicTask periodicTask = PeriodicTaskManager.PeriodicTask.getInstance(beeper,
+ *                 0, 1, TimeUnit.SECONDS, PeriodicTaskManager.class,
+ *                 "Beeper");
+ *         ScheduledFuture<?> future = PeriodicTaskManager.INSTANCE.executePeriodicTask(periodicTask);
+ *
+ *         // Cancel the PeriodicTask
+ *         PeriodicTaskManager.INSTANCE.cancelPeriodicTask(periodicTask);
+ *     }
+ * }}</pre>
+ * @since 2.4.0
+ */
+public final class PeriodicTaskManager {
+
+    /**
+     * A Map which stores the currently active PeriodicTasks and it's associate future.
+     */
+    private final Map<PeriodicTask, ScheduledFuture<?>> periodicTaskMap;
+
+    /**
+     * The executor service which is responsible for running the tasks
+     */
+    private final ScheduledExecutorService periodicTaskService;
+
+    /**
+     * Identifier for the executor service
+     */
+    private final String TASK_POOL_NAME = "AI-SDK-PeriodicTaskManager-Pool";
+
+    /**
+     * Number of threads in the TheadPool associated with periodicTaskService.
+     */
+    private final int poolSize;
+
+    /**
+     * The INSTANCE of {@link PeriodicTaskManager} class
+     */
+    public static PeriodicTaskManager INSTANCE = null;
+
+    private PeriodicTaskManager(int poolSize) {
+        this.poolSize = poolSize;
+        this.periodicTaskService = new ScheduledThreadPoolExecutor(this.poolSize,
+                ThreadPoolUtils.createDaemonThreadFactory(PeriodicTaskManager.class, TASK_POOL_NAME));
+        this.periodicTaskMap = new ConcurrentHashMap<>();
+        SDKShutdownActivity.INSTANCE.register(this.periodicTaskService);
+    }
+
+    /**
+     * Initializer used to initialize the INSTANCE of {@link PeriodicTaskManager}
+     * @param poolSize
+     */
+    public static void initializer(int poolSize) {
+
+        if (poolSize < 1) {
+            throw new IllegalArgumentException("ThreadPool size should be at least 1.");
+        }
+        if (INSTANCE == null) {
+            synchronized (PeriodicTaskManager.class) {
+                if (INSTANCE == null) {
+                    INSTANCE = new PeriodicTaskManager(poolSize);
+                }
+            }
+        }
+    }
+
+    /**
+     * Executes a {@link PeriodicTask}
+     * @param task PeriodicTask
+     * @return ScheduledFuture associated with the scheduled task.
+     */
+    public ScheduledFuture<?> executePeriodicTask(PeriodicTask task) {
+
+        if (task == null) {
+            throw new NullPointerException(" Task cannot be null");
+        }
+
+        task.validate();
```

Comment:
@littleaj I agree with this point. Though, it limits the design of `validate()` method. I expect more parameters to be there in the PeriodicTask in future. Having the `validate()` method bound to parameter will force us to change the signature of the method or write overloaded method. Anyway since this is public, and the the point of creating objects which cannot be used is more costly I have taken your point :)